File: convex/convex/elevenlabs.ts
// ElevenLabs low-latency TTS. Each violation's rationale (and the final
// rewrite) is synthesised into an audio segment, stored in Convex file
// storage, and its storage key pushed onto the session — the UI plays the
// clips in order while the strikethroughs animate. Without a key, the
// frontend falls back to the browser speech engine on the same script.
"use node";
import { v } from "convex/values";
import { action } from "./_generated/server";
import { internal } from "./_generated/api";

export function voiceAvailable(): boolean {
  return !!process.env.ELEVENLABS_API_KEY;
}

// Synthesise one line, store it, attach to the session. Returns the storage id
// (or null when unconfigured — the UI then uses browser speech).
export const speakSegment = action({
  args: { sessionId: v.id("complianceSessions"), text: v.string() },
  handler: async (ctx, { sessionId, text }): Promise<string | null> => {
    const apiKey = process.env.ELEVENLABS_API_KEY;
    if (!apiKey) return null;
    const voice = process.env.ELEVENLABS_VOICE_ID ?? "JBFqnCBsd6RMkjVDRZzb";
    const res = await fetch(`https://api.elevenlabs.io/v1/text-to-speech/${voice}`, {
      method: "POST",
      headers: { "xi-api-key": apiKey, accept: "audio/mpeg", "content-type": "application/json" },
      body: JSON.stringify({
        text,
        model_id: process.env.ELEVENLABS_MODEL ?? "eleven_turbo_v2_5",
        voice_settings: { stability: 0.4, similarity_boost: 0.75, style: 0.15 },
      }),
    });
    if (!res.ok) return null;
    const blob = await res.blob();
    const storageId = await ctx.storage.store(blob);
    await ctx.runMutation(internal.sessions.addVoiceClip, { sessionId, storageKey: storageId });
    return storageId;
  },
});

File: backend/app/services/voice/elevenlabs.py
"""ElevenLabs low-latency TTS for the voice copilot.

available() lets the API tell the frontend whether to stream natural ElevenLabs
audio or fall back to the browser's Web Speech API on the same script — so the
voice demo works today and upgrades to ElevenLabs the moment a key is set.
"""

import httpx

from app.core.config import get_settings


def available() -> bool:
    return bool(get_settings().elevenlabs_api_key)


async def synthesize(text: str) -> bytes:
    """Return MP3 audio for the given text via ElevenLabs streaming TTS."""
    settings = get_settings()
    if not settings.elevenlabs_api_key:
        raise RuntimeError("ELEVENLABS_API_KEY not configured")

    url = f"https://api.elevenlabs.io/v1/text-to-speech/{settings.elevenlabs_voice_id}"
    async with httpx.AsyncClient(timeout=60) as client:
        resp = await client.post(
            url,
            headers={
                "xi-api-key": settings.elevenlabs_api_key,
                "accept": "audio/mpeg",
                "content-type": "application/json",
            },
            json={
                "text": text,
                "model_id": settings.elevenlabs_model,
                "voice_settings": {"stability": 0.4, "similarity_boost": 0.75, "style": 0.15},
            },
        )
        resp.raise_for_status()
        return resp.content
